extra/mariabackup/xtrabackup.cc

```diff
@@ -3589,18 +3590,9 @@ static bool xtrabackup_copy_logfile()
           msg("Error: write to ib_logfile0 failed");
           return true;
         }
-        else
-        {
-          const auto ofs= recv_sys.offset & ~block_size_1;
-          memmove_aligned<64>(log_sys.buf, log_sys.buf + ofs,
-                              recv_sys.len - ofs);
-          recv_sys.len-= ofs;
-          recv_sys.offset&= block_size_1;
-        }
```

Comment:
Neither the function name `xtrabackup_copy_logfile()` nor the removal of this shifting of `log_sys.buf` contents is mentioned in any of the commit messages. What would happen if we have a huge mini-transaction of size `recv_sys.MTR_SIZE_MAX` is starting near the end of the parsing buffer that happens to be of the same size? In this kind of a scenario, we’d want to preserve the already read part of the buffer that contains the start of the mini-transaction, and read the remaining part of the mini-transaction to the rest of the buffer.

Note: I agree that this piece of code may be misplaced here. We’d only need it for handling the case `r == recv_sys_t::PREMATURE_EOF`, a little later in this function.

Reply:
I now realize that unlike in crash recovery, in `mariadb-backup --backup` we must assume that the data that had been previously read into `recv_sys.buf` may be stale. If the log parser reports `recv_sys_t::PREMATURE_EOF`, we cannot know how where we must start re-reading. We must start it from the block that contains the start of the mini-transaction that we are interested in.

It would be good to explain this in the final commit message.